lzy/scheduler/src/main/java/ai/lzy/scheduler/SchedulerApiImpl.java

```diff
@@ -1,125 +1,289 @@
 package ai.lzy.scheduler;
 
+import ai.lzy.scheduler.db.JobsOperationDao;
+import ai.lzy.scheduler.providers.JobSerializer;
+import ai.lzy.longrunning.IdempotencyUtils;
+import ai.lzy.longrunning.Operation;
+import ai.lzy.model.ReturnCodes;
 import ai.lzy.model.TaskDesc;
-import ai.lzy.scheduler.task.Task;
-import ai.lzy.scheduler.worker.Scheduler;
+import ai.lzy.model.db.DbHelper;
+import ai.lzy.scheduler.db.SchedulerDataSource;
+import ai.lzy.scheduler.db.TaskDao;
+import ai.lzy.scheduler.jobs.Allocate;
+import ai.lzy.scheduler.models.TaskState;
+import ai.lzy.util.grpc.JsonUtils;
 import ai.lzy.v1.scheduler.Scheduler.TaskStatus;
 import ai.lzy.v1.scheduler.SchedulerApi.*;
 import ai.lzy.v1.scheduler.SchedulerGrpc;
-import io.grpc.StatusException;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.grpc.Status;
 import io.grpc.stub.StreamObserver;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
+import java.sql.SQLException;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.UUID;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
 @Singleton
 public class SchedulerApiImpl extends SchedulerGrpc.SchedulerImplBase {
-    private final Scheduler scheduler;
+    private static final Logger LOG = LogManager.getLogger(SchedulerApiImpl.class);
+    private final TaskDao dao;
+    private final JobsOperationDao opDao;
+    private final SchedulerDataSource storage;
+    private final Allocate allocateJob;
 
     @Inject
-    public SchedulerApiImpl(Scheduler scheduler) {
-        this.scheduler = scheduler;
-        scheduler.start();
+    public SchedulerApiImpl(TaskDao dao, JobsOperationDao opDao, SchedulerDataSource storage, Allocate allocateJob) {
+        this.dao = dao;
+        this.opDao = opDao;
+        this.storage = storage;
+        this.allocateJob = allocateJob;
     }
 
     @Override
     public void schedule(TaskScheduleRequest request, StreamObserver<TaskScheduleResponse> responseObserver) {
-        final Task task;
+
+        var idempotencyKey = IdempotencyUtils.getIdempotencyKey(request);
+
+        if (idempotencyKey != null) {
+            final Operation op;
+            try {
+                op = DbHelper.withRetries(LOG, () -> opDao.getByIdempotencyKey(idempotencyKey.token(), null));
```

Comment:
Why don't you use `IdempotencyUtils.loadExistingOp` here?

Reply:
Because its not returning operation here